File: day4/cxx/puzzle2.cxx
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include "lines.hpp"
#include "util.hpp"

using namespace std;

const auto solve()
{
    ifstream f{"../input.txt"};

    auto matches = 0u;

    for (auto & line: lines(f))
    {
        // apply passes the tuple returned by `get_range_pair` as arguments to `overlaps`
        if(apply(overlaps, get_range_pair(line)))
        {
            matches++;
        }
    }

    return matches;
}

int main()
{
    auto solution = solve();
    cout << solution << endl;
}

File: day4/py/puzzle2.py
#!/usr/bin/env python

from util import get_range_pair, overlaps

def solve():
    matches = 0

    with open('../input.txt', 'r') as f:
        for line in f:
            if overlaps(*get_range_pair(line)):
                matches += 1

    return matches

if __name__ == "__main__":
    solution = solve()
    print(solution, )

